src-ui.v3/src/Router.hs

```diff
@@ -0,0 +1,258 @@
+
+instance (Semigroup r, MonadHold t m, Adjustable t m) => Adjustable t (SetRouteT t r m) where
+  runWithReplace a0 a' = SetRouteT $ runWithReplace (coerce a0) $ coerceEvent a'
+  traverseIntMapWithKeyWithAdjust f a0 a' = SetRouteT $ traverseIntMapWithKeyWithAdjust (coerce f) (coerce a0) $ coerce a'
+  traverseDMapWithKeyWithAdjust f a0 a' = SetRouteT $ traverseDMapWithKeyWithAdjust (\k v -> coerce $ f k v) (coerce a0) $ coerce a'
+  traverseDMapWithKeyWithAdjustWithMove f a0 a' = SetRouteT $ traverseDMapWithKeyWithAdjustWithMove (\k v -> coerce $ f k v) (coerce a0) $ coerce a'
+
+instance (Monad m, MonadQuery t vs m) => MonadQuery t vs (SetRouteT t r m) where
+  tellQueryIncremental = lift . tellQueryIncremental
+  askQueryResult = lift askQueryResult
+  queryIncremental = lift . queryIncremental
+
+switchPkgRoute :: Set PkgIdxTs -> URI ->  PkgIdxTs -> Maybe URI 
+switchPkgRoute setPkgIdx oldRoute (PkgIdxTs 0) = Nothing
+switchPkgRoute setPkgIdx oldRoute idxChange =
+  let routeS  = (T.pack . uriFragment) oldRoute
+      rootURI = "#/package/"
+  in case T.stripPrefix rootURI routeS of
+    Just sfx | (Just pkgN, Just pkgIdx) <- pkgNFromText sfx
+             , True <- idxChange /= pkgIdx
+             , True <- not (Set.null setPkgIdx)
+             , Just setMax <- Set.lookupMax setPkgIdx
+               -> if setMax == idxChange
+                  then Nothing
+                  else parseURI . T.unpack $ rootURI <> (pkgNToText pkgN) <> (T.pack "@") <> (idxTsToText idxChange)
+             | otherwise -> Nothing
+    Nothing  -> Nothing
+
+routeLink :: forall t m a route. ( DomBuilder t m, SetRoute t (NonEmpty FragRoute) m) 
+          => Bool -- PreventDefault?
+          -> Text -- Target route
+          -> m a -- Child widget
+          -> m a
+routeLink True r w = do
+  let cfg = (def :: ElementConfig EventResult t (DomBuilderSpace m))
+        & elementConfig_eventSpec %~ addEventSpecFlags (Proxy :: Proxy (DomBuilderSpace m)) Click (\_ -> preventDefault)
+        & elementConfig_initialAttributes .~ "href" =: r
+  (e, a) <- element "a" cfg w
+  setRoute $ (((decodeFrag r) NE.:| []) <$ domEvent Click e) 
+  return a
+routeLink False r w = do
+  let cfg = (def :: ElementConfig EventResult t (DomBuilderSpace m))
+        & elementConfig_initialAttributes .~ "href" =: r
+  (e, a) <- element "a" cfg w
+  setRoute $ (((decodeFrag r) NE.:| []) <$ domEvent Click e) 
+  return a
+
+routePkgIdxTs :: forall t m. ( PostBuild t m, MonadHold t m, MonadFix m, DomBuilder t m, SetRoute t (NonEmpty FragRoute) m) 
+              => PkgN
+              -> PkgIdxTs
+              -> Behavior t (Set PkgIdxTs)
+              -> Dynamic t (Map.Map PkgIdxTs Text) 
+              -> DropdownConfig t PkgIdxTs
+              -> m (Dropdown t PkgIdxTs)
+routePkgIdxTs pn k0 setIdx opt ddConf = do
+  dd <- dropdown k0 opt ddConf
+  let evDD = attach setIdx (updated $ dd ^. dropdown_value)
+  setRoute $ ((\val -> (createRoutePackage pn val) NE.:| []) <$> evDD)
```

Comment:
Why is this setRoute here? It's causes duplicate history items that make the back button impossible to use because you always get a new history item once the getPackagesReports gets returned. See https://www.youtube.com/watch?v=zNRvPVXWXxw&feature=youtu.be.